src/graph/network/socket_communicator.cc

```diff
@@ -160,54 +289,31 @@ void SocketReceiver::MsgHandler(TCPSocket* socket, MessageQueue* queue, int id)
       int64_t tmp = socket->Receive(
         reinterpret_cast<char*>(&data_size)+received_bytes,
         max_len);
+      if (tmp == -1) {
+        LOG(FATAL) << "Socket recv error.";
+      }
       received_bytes += tmp;
     }
-    // Data_size ==-99 is a special signal to tell
-    // the MsgHandler to exit the loop
-    if (data_size <= 0) {
-      queue->Signal(id);
-      break;
-    }
-    // Then recv the data
-    received_bytes = 0;
-    while (received_bytes < data_size) {
-      int64_t max_len = data_size - received_bytes;
-      int64_t tmp = socket->Receive(buffer+received_bytes, max_len);
-      received_bytes += tmp;
-    }
-    queue->Add(buffer, data_size);
-  }
-  delete [] buffer;
-}
-
-int64_t SocketReceiver::Recv(char* dest, int64_t max_size) {
-  // Get message from message queue
-  return queue_->Remove(dest, max_size);
-}
-
-void SocketReceiver::Finalize() {
-  for (int i = 0; i <= num_sender_; ++i) {
-    if (i != 0) {  // write -99 signal to exit loop
-      int64_t data_size = -99;
-      queue_->Add(
-        reinterpret_cast<char*>(&data_size),
-        sizeof(int64_t));
-    }
-    if (socket_[i] != nullptr) {
-      socket_[i]->Close();
-      delete socket_[i];
-      socket_[i] = nullptr;
+    if (data_size < 0) {
+      LOG(FATAL) << "Recv data error (data_size: " << data_size << ")";
+    } else if (data_size == 0) {  // This is a end-signal sent by client
+      return;
+    } else {
+      char* buffer = new char[data_size];
```

Comment:
the memory for receiving data is allocated here. but when is it deallocated?

Reply:
For receiver side, the buffer should be allocated by the communicator. Then the ownership should be transferred to the user side. Take NDArray as an example:
1. The buffer allocated in communicator and stored in the message queue.
2. User calls `Recv` and get the data buffer (communicator no longer owns the memory).
3. User creates a DLTensor using the data buffer and correctly set the deallocator.
4. User creates an NDArray from that DLTensor. NDArray is equipped with reference counting, so the data buffer will be destroyed once the NDArray has no reference.